packages/cli/src/parallel-tests/parallel-tests.handler.ts

```diff
@@ -207,35 +123,26 @@ export const runAllTestsInParallel: (
             ? 0
             : resultsForTestCase.numberOfRetriesExecuted + 1;
 
+        resultsByTestId.set(id, {
+          currentResult: mergedResult,
+          numberOfRetriesExecuted,
+        });
+
         // Our work is done for this test case if the first result was a pass,
         // we've performed all the retries, or one of the retries already proved
         // the result as flakey
         const isFinalResult =
           mergedResult.result !== "fail" ||
           numberOfRetriesExecuted === maxRetriesOnFailure;
 
-        resultsByTestId.set(id, {
-          currentResult: mergedResult,
-          numberOfRetriesExecuted,
-        });
-
         if (isFinalResult) {
           // Cancel any replays that are still scheduled
-          queue = queue.filter(({ id }) => id !== id);
+          queue = queue.filter((otherReplay) => otherReplay.id !== id);
 
           finalResults.push(mergedResult);
-          --progress.runningTestCases;
-          progress.failedTestCases += mergedResult.result === "fail" ? 1 : 0;
-          progress.flakedTestCases += mergedResult.result === "flake" ? 1 : 0;
-          progress.passedTestCases += mergedResult.result === "pass" ? 1 : 0;
-          await onTestFinished?.(progress, finalResults).then(() => {
-            if (queue.length === 0 && inProgress === 0) {
-              allTasksDone.resolve();
-            }
-          });
+          progress = updateProgress(progress, mergedResult.result);
```

Comment:
Previously I was mutating progress, which didn't cause any issues, but is generally bad since it could give confusing results if e.g. a onTestFinished call back waited for some promise before reading from the progress object (may get a different result to reading from it before waiting for the promise)